pkg/resolvers/k8s/watcher.go

```diff
@@ -39,63 +54,100 @@ func startNewWatcher(
 	watcherErrs prometheus.Counter,
 	watcherGotChanges prometheus.Counter,
 ) (*watcher, error) {
-	// NOTE(bplotka): Would love to have proper context from above but naming.Resolver does not allow that.
 	ctx, cancel := context.WithCancel(context.Background())
-
-	s, err := startNewStreamer(ctx, target, epClient)
-	if err != nil {
-		return nil, err
-	}
-
 	return &watcher{
-		logger:     logger,
-		ctx:        ctx,
-		cancel:     cancel,
-		target:     target,
-		streamer:   s,
-		addrsState: map[string]struct{}{},
-
+		logger:            logger,
+		ctx:               ctx,
+		cancel:            cancel,
+		target:            target,
+		epClient:          epClient,
+		addrsState:        map[string]struct{}{},
 		resolvedAddrs:     resolvedAddrs,
 		watcherErrs:       watcherErrs,
 		watcherGotChanges: watcherGotChanges,
+		streamer:          nil,
+		streamRetryBackoff: &backoff.Backoff{
+			Min:    50 * time.Millisecond,
+			Jitter: true,
+			Factor: 2,
+			Max:    2 * time.Second,
+		},
 	}, nil
 }
 
 // Close closes the watcher, cleaning up any open connections.
 func (w *watcher) Close() {
+	if w.streamer != nil {
+		w.streamer.Close()
+	}
+
 	w.cancel()
 }
 
+func (w *watcher) startNewStreamerWithRetry(ctx context.Context) *streamer {
+	for ctx.Err() == nil {
+		s, err := startNewStreamer(w.target, w.epClient)
+		if err != nil {
+			s.Close()
+			w.logger.WithError(err).Warnf("k8sresolver: failed to start watching endpoint events for target %v", w.target)
+			time.Sleep(w.streamRetryBackoff.Duration())
+			continue
+		}
+		w.streamRetryBackoff.Reset()
+
+		return s
+	}
+
+	return nil
+}
+
 // Next updates the endpoints for the targetEntry being watched.
 // As from Watcher interface: It should return an error if and only if Watcher cannot recover.
 func (w *watcher) Next() ([]*naming.Update, error) {
 	if w.ctx.Err() != nil {
 		// We already stopped.
-		return []*naming.Update(nil), errors.Wrap(w.ctx.Err(), "k8sresolver: watcher.Next already stopped or Next returned error already. "+
+		return nil, errors.Wrap(w.ctx.Err(), "k8sresolver: watcher.Next already stopped or Next returned error already. "+
 			"Note that watcher errors are not recoverable.")
 	}
-	u, err := w.next()
-	if err != nil {
-		if w.ctx.Err() == nil {
-			// Only update those if watcher is not cancelled.
+
+	ctx, cancel := context.WithCancel(w.ctx)
```

Comment:
What is the point of this context.WithCancel?

Reply:
None - good point